### ISSUE TITLE
BART training: Dynamic shape triggers XLA-compilation on TPU

### ISSUE THREAD

**Issue by JunjieHu** [pytorch/fairseq#2622]
## ❓ Questions and Help
Hi @myleott @kevinmtian @yinhanliu,

I have two questions on training BART with a denosing task on TPU v3-8. I notice a lot of XLA compilations at each update step which cause the training to be very slow. 

1. I read @myleott 's reply on [issue 2324](https://github.com/pytorch/fairseq/issues/2324#issuecomment-658263698). I set `--sample-break-mode=none`, `--tokens-per-sample=64`, and `--max-sentences=16`. Based on my understanding, this would produce a mini-batch of a fixed shape `[16, 64]`. But why would the training still trigger the xla-compilations `INFO | root | NOTE: XLA compilation detected; too many of these can lead to slow training, but we expect a few in the beginning`?  @jiaaoc asked a similar question on [issue 2382](https://github.com/pytorch/fairseq/issues/2382), but the issue seems not resolved yet. 

2. Is it possible that we could create all the mini-batches with a fixed shape (e.g. `[max-sentences, tokens-per-sample]`) before training rather than doing masking on-the-fly, like [the BERT repo](https://github.com/google-research/bert/blob/master/create_pretraining_data.py)? Could you point me to any references about creating pre-training data before hand?

Here are my comments
```
REPO=/home/junjie_j_hu/cap
DATA=$REPO/data/cnn_dm-bin
OUTDIR=$REPO/outputs/bart-cnn-dm
mkdir -p $OUTDIR

python $REPO/fairseq/train.py \
$DATA \
--tpu \
--distributed-world-size=8 \
--max-sentences=16 \
--tokens-per-sample=64 \
--sample-break-mode=none \
--task=denoising \
--save-interval=50 \
--arch=bart_base \
--attention-dropout=0.1 \
--no-progress-bar \
--criterion=cross_entropy \
--lr-scheduler=polynomial_decay \
--min-lr=-1 \
--skip-invalid-size-inputs-valid-test \
--update-freq=[2] \
--optimizer=adam \
--adam-betas="(0.9, 0.98)" \
--lr=[0.0006] \
--warmup-updates=10000 \
--share-decoder-input-output-embed \
--dropout=0.1 \
--weight-decay=0.01 \
--train-subset=train \
--valid-subset=valid \
--max-update=500000 \
--save-dir=$OUTDIR/ \
--restore-file=checkpoint_last.pt \
--mask=0.3 \
--mask-random=0.1 \
--poisson-lambda=3.5 \
--permute-sentences=1 \
--mask-length=span-poisson \
--replace-length=1 \
--max-source-positions=64 \
--max-target-positions=64 \
--share-all-embeddings \
--layernorm-embedding \
--log-interval=10 \
--log-format=json \
--tensorboard-logdir=$OUTDIR/logs \
--seed=1 \
--min-loss-scale=0.0001 \
--model-parallel-size=1 \
--num-workers=4 \
--required-batch-size-multiple=1 \
--validate-interval=1 \
--bucket-cap-mb=25 \
--clip-norm=0.1 \
--optimizer-overrides={} \
--save-interval-updates=50 \
--keep-interval-updates=-1 \
--keep-last-epochs=-1 \
--keep-best-checkpoints=-1 \
--no-epoch-checkpoints \
--best-checkpoint-metric=loss \
--patience=-1 \
--adam-eps=1e-06 \
--power=1 \
--total-num-update=100
```
Here are the full logs
```2020-09-16 06:50:48 | INFO | fairseq_cli.train | task: denoising (DenoisingTask)
2020-09-16 06:50:48 | INFO | fairseq_cli.train | model: bart_base (BARTModel)
2020-09-16 06:50:48 | INFO | fairseq_cli.train | criterion: cross_entropy (CrossEntropyCriterion)
2020-09-16 06:50:48 | INFO | fairseq_cli.train | num. model params: 137945856 (num. trained: 137945856)
2020-09-16 06:50:53 | INFO | fairseq.trainer | detected shared parameter: encoder.embed_tokens.weight <- decoder.embed_tokens.weight
2020-09-16 06:50:53 | INFO | fairseq.trainer | detected shared parameter: encoder.embed_tokens.weight <- decoder.output_projection.weight
2020-09-16 06:50:53 | INFO | fairseq_cli.train | training on 8 devices (GPUs/TPUs)
2020-09-16 06:50:53 | INFO | fairseq_cli.train | max tokens per GPU = None and max sentences per GPU = 16
2020-09-16 06:50:53 | INFO | fairseq.trainer | no existing checkpoint found /home/junjie_j_hu/cap/outputs/bart-cnn-dm-v4/checkpoint_last.pt
2020-09-16 06:50:53 | INFO | fairseq.trainer | loading train data for epoch 1
2020-09-16 06:50:53 | INFO | fairseq.data.data_utils | loaded 287227 examples from: /home/junjie_j_hu/cap/data/cnn_dm-bin/train
2020-09-16 06:50:53 | INFO | fairseq.tasks.denoising | Split: train, Loaded 4024307 samples of denoising_dataset
2020-09-16 06:51:12 | INFO | fairseq.trainer | begin training epoch 1
2020-09-16 06:52:28 | INFO | root | NOTE: XLA compilation detected; too many of these can lead to slow training, but we expect a few in the beginning
2020-09-16 06:53:37 | INFO | root | NOTE: XLA compilation detected; too many of these can lead to slow training, but we expect a few in the beginning
2020-09-16 06:54:41 | INFO | root | NOTE: XLA compilation detected; too many of these can lead to slow training, but we expect a few in the beginning
2020-09-16 06:55:24 | INFO | root | NOTE: XLA compilation detected; too many of these can lead to slow training, but we expect a few in the beginning
2020-09-16 06:55:47 | INFO | root | NOTE: XLA compilation detected; too many of these can lead to slow training, but we expect a few in the beginning
2020-09-16 06:56:09 | INFO | root | NOTE: XLA compilation detected; too many of these can lead to slow training, but we expect a few in the beginning
2020-09-16 06:56:55 | INFO | root | NOTE: XLA compilation detected; too many of these can lead to slow training, but we expect a few in the beginning
2020-09-16 06:57:39 | INFO | root | NOTE: XLA compilation detected; too many of these can lead to slow training, but we expect a few in the beginning
2020-09-16 06:57:39 | INFO | train_inner | {"epoch": 1, "update": 0.001, "loss": "15.861", "ppl": "59509.1", "wps": "0", "ups": "0", "wpb": "16384", "bsz": "256", "num_updates": "10", "lr": "6e-07", "gnorm": "2.138", "clip": "100", "train_wall": "370", "wall": "406"}
```

#### What's your environment?

 - fairseq Version: 0.9.0
 - PyTorch Version:  pytorch-xla-1.6
 - OS (e.g., Linux): Debian
 - TPU v3-8
